ydb/core/tx/datashard/prefix_kmeans.cpp

```diff
@@ -409,82 +408,59 @@ class TPrefixKMeansScan final: public TPrefixKMeansScanBase, private TCalculatio
     {
         this->Dimensions = request.GetSettings().vector_dimension();
-        LOG_D("Create " << Debug());
+        LOG_I("Create " << Debug());
     }
 
     EScan Seek(TLead& lead, ui64 seq) final
     {
-        LOG_D("Seek " << Debug());
-        ui64 zeroSeq = 0;
-        while (true) {
-            if (State == UploadState) {
-                if (MoveToNextPrefix()) {
-                    zeroSeq = seq;
-                    continue;
-                }
-                
-                if (UploadInProgress()
-                    || TryUpload(LevelBuf, LevelTable, LevelTypes, false)
-                    || TryUpload(PostingBuf, PostingTable, PostingTypes, false)
-                    || TryUpload(PrefixBuf, PrefixTable, PrefixTypes, false))
-                {
-                    return EScan::Sleep;
-                }
-
-                return EScan::Final;
+        LOG_D("Seek " << seq << " " << Debug());
+
+        if (IsExhausted) {
+            if (UploadInProgress()
+                || TryUpload(LevelBuf, LevelTable, LevelTypes, false)
+                || TryUpload(PostingBuf, PostingTable, PostingTypes, false)
+                || TryUpload(PrefixBuf, PrefixTable, PrefixTypes, false))
+            {
+                return EScan::Sleep;
             }
+            return EScan::Final;
+        }
 
-            lead = Lead;
-            if (State == EState::SAMPLE) {
-                lead.SetTags({&EmbeddingTag, 1});
-                if (seq == zeroSeq && !HasNextPrefix) {
-                    return EScan::Feed;
-                }
-                State = EState::KMEANS;
-                if (!InitAggregatedClusters()) {
-                    // We don't need to do anything,
-                    // because this datashard doesn't have valid embeddings for this parent
-                    if (MoveToNextPrefix()) {
-                        zeroSeq = seq;
-                        continue;
-                    }
-                    return EScan::Final;
-                }
-                ++Round;
-                return EScan::Feed;
-            }
+        lead = Lead;
 
-            Y_ASSERT(State == EState::KMEANS);
-            if (RecomputeClusters()) {
-                lead.SetTags(UploadScan);
-                WriteLevelRows();
-                State = UploadState;
-            } else {
-                lead.SetTags({&EmbeddingTag, 1});
-                ++Round;
-            }
-            return EScan::Feed;
-        }
+        return EScan::Feed;
     }
 
     EScan Feed(TArrayRef<const TCell> key, const TRow& row) final
     {
         LOG_T("Feed " << Debug());
+
+        ++ReadRows;
+        ReadBytes += CountBytes(key, row);
+
+        if (Prefix && !TCellVectorsEquals{}(Prefix.GetCells(), key.subspan(0, PrefixColumns))) {
+            if (!FinishPrefix()) {
+                // scan current prefix rows with a new state again
+                return EScan::Reset;
+            }
+        }
+
         if (!Prefix) {
```

Comment:
this code is needed only for FeedSample, we don't need it if it's not `IsFirstPrefixFeed`